### PR TITLE
[CALCITE-2812] Add algebraic operators to allow expressing recursive queries (Ruben Quesada Lopez)

### PR DESCRIPTION
**Pull request by rubenada**
New operator RepeatUnion (implementation based on the new entities TableSpool and TransientTable).
New unit tests containing some basic examples.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/jdbc/CalciteConnectionImpl.java [hunk 1]
```diff
@@ -420,6 +421,7 @@ public CalciteServerStatement getStatement(Meta.StatementHandle h)
           .put(Variable.TIME_ZONE.camelName, timeZone)
           .put(Variable.STDIN.camelName, streamHolder.get()[0])
           .put(Variable.STDOUT.camelName, streamHolder.get()[1])
+          .put("#DELTA#", new LinkedList<>()) // TODO rql cleaner way?
```
Comment (zabetak) [inline core/src/main/java/org/apache/calcite/jdbc/CalciteConnectionImpl.java]: I guess this can be done inside EnumerableDeltaTableScan#implement, no?

Comment (rubenada) [reply core/src/main/java/org/apache/calcite/jdbc/CalciteConnectionImpl.java]: I think when we arrive there it is "too late": at that point, we can only read from the DataContext via 'get' method, but we cannot put new parameters in there.

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/jdbc/CalciteConnectionImpl.java]: What if we have recursive (recursive) query?
Then a single DELTA would be not enough.

By the way: please never ever use `LinkedList`.
Here's what the author says: https://twitter.com/joshbloch/status/583813919019573248

Comment (zabetak) [reply core/src/main/java/org/apache/calcite/jdbc/CalciteConnectionImpl.java]: Another alternative would be to add a new Table during EnumerableDeltaTableScan#implement using DataContext#getRootSchema. This can solve also the problem of having multiple levels of recursion since each different scan can establish a different table name. How about something in these lines?

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/jdbc/CalciteConnectionImpl.java]: Is it really required to add a table?

For instance, EnumerableJoin manages to operate yet it doesn't require adding a table for that.

Comment (zabetak) [reply core/src/main/java/org/apache/calcite/jdbc/CalciteConnectionImpl.java]: In this case data have to shared/transferred between different operators (i.e., Union + Scan) which do not necessarily have parent-child relationships. Due to this I would say that it is necessary to introduce "something" that can be shared between operators and I think that a Table would be a good abstraction but maybe I am missing something.

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/jdbc/CalciteConnectionImpl.java]: @zabetak , I wonder if we should add a new `RelNode` subtype for "recursive with delta" table instead.

Adding tables to the schema for a sake of a single query on the fly does not seem right.

Comment (rubenada) [reply core/src/main/java/org/apache/calcite/jdbc/CalciteConnectionImpl.java]: The current solution is a naive one, which, as you say, does not support multiple recursive declarations.
In order to address this issue, and following the current approach, we could create each deltaScan with a certain parameter to provide a unique identifier (what about a ModifiableTable?), and then this parameter will have to be transferred also in the moment of the recursiveUnion creation, so that it can be aware of "where to put" the result data (so that it can be picked by the deltaScan in the next iteration).

Comment (hsyuan) [reply core/src/main/java/org/apache/calcite/jdbc/CalciteConnectionImpl.java]: We don't really need an additional table. In Postgres' implementation, RecursiveUnion and DeltaTableScan both have a pointer pointing to each other. DeltaTableScan stores the intermediate data store that RecursiveUnion can access.

Comment (rubenada) [reply core/src/main/java/org/apache/calcite/jdbc/CalciteConnectionImpl.java]: Committed new version:
- deltaTableScan and recursiveUnion share an identifier (the shared delta table) to carry out the data transfer, this allows to have multiple (nested) recursive unions, unit tests provided showing this scenario.
- implemented (although not used yet) new parameter maxDepth in deltaTableScan (might be useful for graph traversal)

Still TODO: 
- Data transfer between deltaTableScan and recursiveUnion is still done via a parameter in the DataContext, this has to be modified. I'll explorer the approach proposed by @zabetak of using the SchemaPlus via DataContext#getRootSchema

Comment (rubenada) [reply core/src/main/java/org/apache/calcite/jdbc/CalciteConnectionImpl.java]: Committed a new version that address the TODO mentioned in the comment above: instead of using the DataContext, now the data store / transfer between recursiveUnion and deltaTableScan is done via a transient table (a simple ModifiableTable) that gets added to the schema.

Comment (zabetak) [reply core/src/main/java/org/apache/calcite/jdbc/CalciteConnectionImpl.java]: Thanks @rubenada! I like the new approach. Let's see what @vlsi , and @hsyuan have to say about this.

File: core/src/test/java/org/apache/calcite/test/enumerable/EnumerableRecursiveUnionTest.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+import java.util.Arrays;
+
+/**
+ * Unit tests for
+ * {@link org.apache.calcite.adapter.enumerable.EnumerableRecursiveUnion}
+ * and {@link org.apache.calcite.adapter.enumerable.EnumerableDeltaTableScan}
+ * <a href="https://issues.apache.org/jira/browse/CALCITE-2812">[CALCITE-2812]
+ * Add algebraic operators to allow expressing recursive queries</a>.
+ */
+public class EnumerableRecursiveUnionTest {
```
Comment (zabetak) [inline core/src/test/java/org/apache/calcite/test/enumerable/EnumerableRecursiveUnionTest.java]: @rubenada can you add also a few tests where the recursive part is a bit more complicated (for instance a join between a delta scan and a regular table)?

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java [hunk 3]
```diff
@@ -3329,6 +3329,167 @@ public void reset() {
...
+                this.currentAddedToDelta = false;
+                this.current = this.seedEnumerator.current();
+                return true;
+              } else {
+                this.seedProcessed = true;
+              }
+            }
+
+            // if we are done with the seed, moveNext on the recursive part
+            if (this.recursionEnumerator.moveNext()) {
```
Comment (zabetak) [inline linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: From this line it seems that the recursive part can contain only certain kind of operators. It seems that all operators above the delta scan table must allow pipelined execution (i.e., read rows one by one). If for some reason there is an operator that reads all rows before producing any result then probably the result of the recursive union may not be correct. 

If what I described above is true then we could try the alternative below.
When `this.recursionEnumerator.moveNext()` returns false then we obtain a new enumerator and keep repeating this procedure till there are not really any new rows (e.g., the new enumerator returns directly false?).
If `maxDepth` was part of the recursive union then we could stop after obtaining a new enumerator a certain number of times. 

What do you think @rubenada ?

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java [hunk 4]
```diff
@@ -1616,6 +1619,68 @@ public RelBuilder minus(boolean all, int n) {
...
+  public RelBuilder deltaScan(String tableName) {
+    return this.deltaScan(tableName, this.peek().getRowType());
+  }
+
+  /**
+   * Creates a {@link LogicalDeltaTableScan} of a delta work table (used to accumulate results in
+   * recursive union operation), using as its row type the top of the stack, and a certain max depth
+   * Returns this builder.
+   */
+  public RelBuilder deltaScan(String tableName, int maxDepth) {
```
[the remark this answers is not in the document]
Comment (rubenada) [reply core/src/main/java/org/apache/calcite/tools/RelBuilder.java]: @zabetak you're right, I have committed a new version with a refactoring using this approach.

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 3 above)
Comment (rubenada) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: @zabetak I agree, I have committed a new version with a refactoring using this approach.

File: core/src/test/java/org/apache/calcite/test/enumerable/EnumerableRecursiveUnionTest.java (hunk 2 above)
Comment (rubenada) [reply core/src/test/java/org/apache/calcite/test/enumerable/EnumerableRecursiveUnionTest.java]: New unit tests have been added

File: core/src/main/java/org/apache/calcite/schema/impl/TransientTable.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * Simple modifiable table whose elements can be accessed / altered via
+ * {@link #getModifiableCollection()} method (the rest of the methods are not supported).
+ * It can be used, for example, in a recursive union to accumulate all intermediate results.
+ * @param <T> table content type
+ */
+public class TransientTable<T> implements ModifiableTable {
```
Comment (zabetak) [inline core/src/main/java/org/apache/calcite/schema/impl/TransientTable.java]: Although I proposed transient table as a name, seeing the implementation/javadoc a name like ListModifiableTable seems more appropriate. 

Other than that the fact that the class is public and does not add anything new to ModifiableTable (plus it throws an exception almost in every case) makes me a bit skeptical. I am not sure if we should really invest on this but it could be nice if could reduce the visibility of this class or come up with a solution where we don't really introduce a new public class (e.g., static factory method returning a ModifiableTable)?

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java [hunk 6]
```diff
@@ -1616,6 +1619,58 @@ public RelBuilder minus(boolean all, int n) {
     return setOp(all, SqlKind.EXCEPT, n);
   }
 
+  /**
+   * Creates a {@link LogicalDeltaTableScan} of a delta work table (used to accumulate results in
+   * recursive union operation), using as its row type the top of the stack.
+   * Returns this builder.
+   */
+  public RelBuilder deltaScan(String tableName) {
```
Comment (zabetak) [inline core/src/main/java/org/apache/calcite/tools/RelBuilder.java]: Since the notion of Delta already exists for the case of streams I would suggest to use the notion of transient tables for the new operators. For example deltaScan becomes transientScan, LogicalDeltaTableScan becomes LogicalTransientTableScan etc.

File: core/src/test/java/org/apache/calcite/test/enumerable/EnumerableRecursiveUnionTest.java [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+      );
+  }
+
+  // Tests for the following hierarchy:
+  //      Emp1
+  //      /  \
+  //    Emp2  Emp4
+  //    /  \
+  // Emp3   Emp5
+  @Test public void testHierarchy() {
```
Comment (zabetak) [inline core/src/test/java/org/apache/calcite/test/enumerable/EnumerableRecursiveUnionTest.java]: The test looks like a good candidate for being parameterized (see classes anotated with @RunWith(Parameterized.class)).

Comment (vlsi) [reply core/src/test/java/org/apache/calcite/test/enumerable/EnumerableRecursiveUnionTest.java]: +1

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableDeltaTableScan.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+import org.apache.calcite.rel.core.TableScan;
+import org.apache.calcite.schema.ModifiableTable;
+import org.apache.calcite.util.BuiltInMethod;
+
+
+/**
+ * Implementation of delta table (used to accumulate results in a recursive query) in
+ * {@link org.apache.calcite.adapter.enumerable.EnumerableConvention enumerable calling convention}.
+ */
+public class EnumerableDeltaTableScan extends TableScan implements EnumerableRel {
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableDeltaTableScan.java]: I wonder if this new subtype of `TableScan` would backfire when non-expecting rules would fire.
I think it would be better to have a standalone rel type.

File: core/src/test/java/org/apache/calcite/test/enumerable/EnumerableRecursiveUnionTest.java (hunk 7 above)
Comment (rubenada) [reply core/src/test/java/org/apache/calcite/test/enumerable/EnumerableRecursiveUnionTest.java]: Refactored

File: core/src/test/java/org/apache/calcite/test/JdbcTest.java [hunk 9]
```diff
@@ -6837,6 +6837,36 @@ public TranslatableTable view(String s) {
...
+            new Department(10, "Dept", Arrays.asList(emps[0], emps[1], emps[2], emps[3], emps[4]),
+                    new Location(-122, 38)),
+    };
+
+    //      Emp1
+    //      /  \
+    //    Emp2  Emp4
+    //    /  \
+    // Emp3   Emp5
+    public final Hierarchy[] hierarchies = {
```
[the remark this answers is not in the document]
Comment (rubenada) [reply core/src/test/java/org/apache/calcite/test/JdbcTest.java]: I think this design has two (small) advantages:
- It can represent more complicated hierarchies, e.g. one employee with more than one manager (although this is currently not used in the test cases)
- It requires 2 joins (instead of 1) to compute the hierarchy paths, so it allows us to show a trickier use case of recursive union + transient table scan

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java (hunk 6 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/tools/RelBuilder.java]: +1
Delta would indeed confuse.
Especially for `LogicalDeltaDeltaScan`

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableDeltaTableScan.java (hunk 8 above)
Comment (zabetak) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableDeltaTableScan.java]: By looking into the code I don't see something fundamentally different from a regular table scan. Due to this I am less worried about non-expecting rules firing. However, I have the feeling that EnumerableDeltaTableScan and likewise LogicalDeltaTableScan can completely dissapear. LogicalDeltaTableScan can be possibly replaced by a regular LogicalTableScan and EnumerableDeltaTableScan by EnumerableTableScan. The logic of obtaining a table from the context could be part of the Table implementation.

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java [hunk 10]
```diff
@@ -1616,6 +1619,58 @@ public RelBuilder minus(boolean all, int n) {
...
+   * @param tableName table name
+   * @param rowType row type of the table
+   */
+  public RelBuilder deltaScan(String tableName, RelDataType rowType) {
+    RelOptTable table = RelOptTableImpl.create(
+            this.relOptSchema,
+            rowType,
+            Collections.singletonList(tableName),
+            null);
+    LogicalDeltaTableScan scan = LogicalDeltaTableScan.create(this.cluster, table);
```
Comment (zabetak) [inline core/src/main/java/org/apache/calcite/tools/RelBuilder.java]: Regarding my previous comment of possibly removing LogicalDeltaTableScan (replacing with LogicalTableScan) I was thinking that the RelOptTableImpl above could be enhanced with adequate information which would make the need for a new scan operator obsolete.

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java (hunk 6 above)
Comment (rubenada) [reply core/src/main/java/org/apache/calcite/tools/RelBuilder.java]: Agree, *delta* renamed as *transient*

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableRecursiveUnion.java [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+import org.apache.calcite.util.BuiltInMethod;
+
+import java.util.List;
+
+
+/**
+ * Implementation of Recursive Union in
+ * {@link EnumerableConvention enumerable calling convention}.
+ */
+public class EnumerableRecursiveUnion extends Union implements EnumerableRel {
```
Comment (zabetak) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableRecursiveUnion.java]: As @vlsi said earlier I am not sure if this new operator (which is a bit particular) should extend Union since there is a risk that existing rules may match when they shouldn't. 

Apart from that I can identify two main functionalities in this operator:
1. Iterative calls to the subplan till a certain number is reached or untill there are no more results.
2. Buffer and write to a side table/collection at certain intervals (i.e., end of iteration).

I think we could try to isolate these functionalities in separate logical/physical operators with clear semantics and responsibilities which could improve encapsulation and reutilization of these operators in other use cases. 

I propose something along the following lines. For 1, we could have a Logical/EnumerableIterate operator while for 2, we could have a Logical/EnumerableSpool (maybe relevant to look at CALCITE-481) that takes care of buffering and writting to the side table. Then to satisfy the use cases so far the two operators should be combined the one on top of the other (Iterate[Spool[Subplan]]). Last thing that is missing is the seed which could be placed somwhere in the subplan and we should ensure somehow that is executed only once (using a filter?).

@rubenada If what I am proposing appears very complicated, useless, or out of context please let me know. In any case the work so far is great :)

File: core/src/test/java/org/apache/calcite/test/JdbcTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (zabetak) [reply core/src/test/java/org/apache/calcite/test/JdbcTest.java]: OK, agreed!

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableRecursiveUnion.java (hunk 11 above)
Comment (rubenada) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableRecursiveUnion.java]: @zabetak Thanks for your feedback. I'll propose the following re-design to implement the "recursive union all" with re-usable modules:

Firstly, Create a new `abstract class Iterate extends BiRel` with its corresponding logical / physical operators `LogicalIterate` and `EnumerableIterate`. This Iterate will have two inputs:
- The left part will be used as `seed`, and will be evaluated once
- The right part will be used as `iteration`, and will be re-evaluated over and over until no results are produced (or until an optional `maxIter` value is reached)
This design will allow us to use the `Iterate` as main implementation module for the recursive union (which requires a non-recursive term, executed once; and a recursive term, executed over and over until no results are returned).
If, in the future, this `Iterate` needs to be used for other purposes that require just a "pure" iteration (no seed needed), it will still be possible to used it: we will just need to create an Iterate with an empty Values as left (i.e. as seed).

Secondly, to implement the data transfer from the recursive union into the "work table", a new, simple `Spool` operator will be created (see CALCITE-481). For the sake of this PR, a naive, lazy spool implementation will be enough. This operator will store in a temporary collection the values returned by its input, before propagating them. When the input returns no more values, the `Spool` will flush its temporary collection into another one (which in the case of the recursive union will be the same collection that will be used to hold the work table).

Finally, the temporary table will be a scan on table represented by a list (which will be re-initialized by the `Spool`).

To sum up, to represent a recursive union we would have an `Iterate`, with two `Spool`s inserted on top of its left (seed) and right (iteration).

To show an example, a typical recursive union like (return numbers from 1 to 10):
```
WITH RECURSIVE delta(n) AS (
    VALUES (1)
    UNION ALL
    SELECT n+1 FROM delta WHERE n < 10
)
SELECT * FROM delta
```
would be represented with the pseudo-plan:
```
Iterate
  Spool(table=delta) -- seed, executed once
    Values({1})
  Spool(table=delta) -- iteration, executed until no results are returned
    Project($0=$0+1)
      Filter($0<10)
        Scan(table=delta)
```

Comment (zabetak) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableRecursiveUnion.java]: I like the new design but before going forward let's see what the others have to say. Just a few additional information for those following the discussion.

There has been a small debate on the name of the RepeatUnion. `Iterate` and `RecursiveUnion` were the other names that were abandoned. 

Although `Iterate`, initially seemed to be a nice abstraction, at the end of the day the description of what it should do was a union of the results of its inputs with the particularity that we always start from the `seedRel` and we repeatedly union the `iterativeRel`.

`RecursiveUnion` also seemed nice, since it is a named also used by Postgres, and fits well with the `RECURSIVE` keyword of SQL. However, it gives the false impression that the operator does some kind of recursion which is not the case.

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java [hunk 12]
```diff
@@ -1705,6 +1711,61 @@ public RelBuilder minus(boolean all, int n) {
     return setOp(all, SqlKind.EXCEPT, n);
   }
 
+  /**
+   * Creates a {@link LogicalTableScan} on a {@link TransientTable} (used e.g. to accumulate
```
Comment (zabetak) [inline core/src/main/java/org/apache/calcite/tools/RelBuilder.java]: It is better if examples are not in the first sentence of the Javadoc.

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java [hunk 13]
```diff
@@ -1705,6 +1711,61 @@ public RelBuilder minus(boolean all, int n) {
...
+   * Creates a {@link LogicalRepeatUnion} of the two most recent relational expressions
+   * on the stack. Warning: if these relational expressions are not correctly defined,
+   * this operation might lead to an infinite loop.
+   * Returns this builder.
+   * @param tableName transient table name associated to the repeat union
+   * @param maxRep maximum number of iterations, -1 means no limit
+   */
+  @Experimental
+  public RelBuilder repeatUnion(String tableName, int maxRep) {
+    RelNode iter = LogicalTableSpool.create(build(), Spool.Type.LAZY, Spool.Type.LAZY, tableName);
```
Comment (zabetak) [inline core/src/main/java/org/apache/calcite/tools/RelBuilder.java]: I guess it is convenient for the end-users to not have to build separately the spool when they want to perform the classical recursive union but this limits a bit the capabilities of the builder. It may be better to provide a separate method spool to do this. Clients who are only interested for this use-case can build the utility methods themselves.

Comment (zabetak) [reply core/src/main/java/org/apache/calcite/tools/RelBuilder.java]: Plus instead of using directly the create method it would be preferable to use factories that are passed through the context to the builder as it is done for the other operators.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableRepeatUnionRule.java [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+  }
+
+  @Override public RelNode convert(RelNode rel) {
+    LogicalRepeatUnion union = (LogicalRepeatUnion) rel;
+    EnumerableConvention out = EnumerableConvention.INSTANCE;
+    RelTraitSet traitSet = union.getTraitSet().replace(out);
+    RelNode seedRel = union.getSeedRel();
+    RelNode iterativeRel = union.getIterativeRel();
+
+    return new EnumerableRepeatUnion(
```
[the remark this answers is not in the document]
Comment (rubenada) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableRepeatUnionRule.java]: This is not the case for other similar operators (EnumerableUnion / Minus / Intersect), that EnumerableRepeatUnion is "inspired" by. I guess because RelTraitSet must be deduce from the LogicalUnion / Minus / Intersect at rule matching time.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableTableSpool.java [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+ * Implementation of {@link Spool} in
+ * {@link EnumerableConvention enumerable calling convention}
+ * that writes into a {@link ModifiableTable}
+ *
+ * <p>NOTE: The current API is experimental and subject to change without notice.</p>
+ */
+@Experimental
+public class EnumerableTableSpool extends Spool implements EnumerableRel {
+
+  // table name, it must refer to an existing ModifiableTable
```
[the remark this answers is not in the document]
Comment (rubenada) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableTableSpool.java]: Comment updated

File: core/src/main/java/org/apache/calcite/rel/core/Spool.java [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelTraitSet;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.RelWriter;
+import org.apache.calcite.rel.SingleRel;
+
+import java.util.List;
+
+/**
+ * Relational expression that consumes a set of data, and stores it in a temporary structure.
```
[the remark this answers is not in the document]
Comment (rubenada) [reply core/src/main/java/org/apache/calcite/rel/core/Spool.java]: Comment modified

File: core/src/main/java/org/apache/calcite/rel/core/Spool.java [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+ *
+ * <p>NOTE: The current API is experimental and subject to change without notice.</p>
+ */
+@Experimental
+public abstract class Spool extends SingleRel {
+
+  /**
+   * Enumeration representing spool read / write type (all four combinations are possible):
```
[the remark this answers is not in the document]
Comment (rubenada) [reply core/src/main/java/org/apache/calcite/rel/core/Spool.java]: Javadoc updated
I think that readType/writeType are more meaningful (and clear) than iterateType/forwardType

File: core/src/main/java/org/apache/calcite/schema/impl/TransientTable.java [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+import org.apache.calcite.schema.SchemaPlus;
+import org.apache.calcite.schema.Schemas;
+
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * Simple modifiable table backed by a Java list.
```
[the remark this answers is not in the document]
Comment (rubenada) [reply core/src/main/java/org/apache/calcite/schema/impl/TransientTable.java]: Class renamed, Javadoc updated.
extend AbstractQueryableTable is needed because interface ModifiableTable extends QueryableTable

File: core/src/main/java/org/apache/calcite/rel/logical/LogicalRepeatUnion.java [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+  /** Creates a LogicalRepeatUnion. */
+  public static LogicalRepeatUnion create(RelNode seed, RelNode iterative) {
+    return create(seed, iterative, -1);
+  }
+
+  /** Creates a LogicalRepeatUnion. */
+  public static LogicalRepeatUnion create(RelNode seed, RelNode iterative, int maxRep) {
+    RelOptCluster cluster = seed.getCluster();
+    RelTraitSet traitSet = cluster.traitSetOf(Convention.NONE);
+    return new LogicalRepeatUnion(maxRep, cluster, traitSet, seed, iterative);
```
[the remark this answers is not in the document]
Comment (rubenada) [reply core/src/main/java/org/apache/calcite/rel/logical/LogicalRepeatUnion.java]: I followed here the same logic as LogicalUnion, I'm not sure we can / should do otherwise

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java (hunk 13 above)
Comment (rubenada) [reply core/src/main/java/org/apache/calcite/tools/RelBuilder.java]: Spool and RepeatUnion factories added
